staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 }
 
 // nodeSyncService syncs the nodes for one load balancer type service
-func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+func (s *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
 	if svc == nil || !wantsLoadBalancer(svc) {
 		return false
 	}
 	klog.V(4).Infof("nodeSyncService started for service %s/%s", svc.Namespace, svc.Name)
-	hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
-	if err != nil {
-		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
-		return true
+	newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(true))
+	// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready
+	// if these hosts are not actually hosting any endpoints for the service. Doing that leads to
+	// unnecessary updates for LBs which will never send traffic to the nodes in question in any case.
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+		oldReadyNodes := filterWithPredicate(oldNodes, s.getNodeConditionPredicate(true))
+		oldReadyNodeNames := nodeNames(oldReadyNodes)
+		newReadyNodeNames := nodeNames(newReadyNodes)
+		newNotReadyNodes := oldReadyNodeNames.Difference(newReadyNodeNames)
+		newReadyNodes := newReadyNodeNames.Difference(oldReadyNodeNames)
+		// We have four cases to consider here:
+		// a) Node ADD
+		// b) Node DEL
+		// [Node UPDATE]
+		//   c) Ready -> NotReady
+		//   d) NotReady -> Ready
+		// We only want to possibly __not__ sync on c) and d).
+		if len(oldNodes) == len(newNodes) {
```

Comment:
I feel like I am missing something.  Why do we care if the lengths match?  Is it possible that I added a node an removed a different node fast enough that this captures both?

It feels like this could be easier to follow, am I missing some finer point?

```
// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready
// if these hosts are not actually hosting any endpoints for the service. Doing that leads to
// unnecessary updates for LBs which will never send traffic to the nodes in question in any case.
if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
	needsSync := func() bool {
		// We have four cases to consider here:
		//   a) Node ADD
		//   b) Node DEL
		//   c) Node UPDATE Ready -> NotReady
		//   d) Node UPDATE NotReady -> Ready
		// In case of (c) or (d) we can skip sync for ETP=Local Service
		oldNodeNames := nodeNames(oldNodes)
		newNodeNames := nodeNames(newNodes)
		if !oldNodeNames.Equal(newNodeNames) {
			// The set of nodes has changed - needs sync.
			return true
		}
		// The set of nodes didn't change, but maybe readiness did.
		oldReadyNodes := filterWithPredicate(oldNodes, s.getNodeConditionPredicate(true))
		newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(true))
		oldReadyNodeNames := nodeNames(oldReadyNodes)
		newReadyNodeNames := nodeNames(newReadyNodes)

		// Figure out which nodes changed readiness.
		changedNodeNames := oldReadyNodeNames.Difference(newReadyNodeNames)
		changedNodeNames = changedNodeNames.Union(newReadyNodeNames.Difference(oldReadyNodeNames))
		if len(changedNodeNames) == 0 {
			// Something else changed - needs sync.
			return true
		}

		if s.serviceHasEndpointsOnNodes(svc, changedNodeNames) {
			// This service has at least 1 endpoint on a changed
			// node - needs sync.
				return true
		}
		// It seems safe to skip this synv for this service.
		return false
	}
	if !needsSync() {
		return false
	}
}
```

Reply:
I put the "Something else changed - needs sync" block in as the sort of "default" case - something I don't understand happened and it seems safer to sync.  We should probably log something useful here, since it should not happen.